logging/writer.go

```diff
@@ -1,29 +1,70 @@
 package logging
 
-import "io"
+import (
+	"bytes"
+	"io"
+	"strings"
+	"sync"
+)
 
 // writer is an implementation of io.Writer that writes to a logger.
 type writer struct {
 	l Logger
 	f func(Logger, string)
+
+	mu  sync.Mutex
+	buf strings.Builder
 }
 
 // NewWriter returns an io.Writer that writes content to the given logger.
 //
-// Each call to Write() produces a separate log message.
+// The string passed to Write() is sliced by end-of-line (EOL) characters and
+// each substring in between the EOL characters is written as a separate log
+// message.
+//
+// If the entire string or its last section does not terminate with an EOL
+// character, it is buffered inside the writer. The writer keeps buffering the
+// string until it detects an EOL character and produces a log message with the
+// buffered string content before the EOL character.
 func NewWriter(l Logger) io.Writer {
-	return &writer{l, LogString}
+	return &writer{l: l, f: LogString}
 }
 
 // NewDebugWriter returns an io.Writer that writes content to the given logger
 // as debug messages.
 //
-// Each call to Write() produces a separate log message.
+// The string passed to Write() is sliced by end-of-line (EOL) characters and
+// each substring in between the EOL characters is written as a separate log
+// message.
+//
+// If the entire string or its last section does not terminate with an EOL
+// character, it is buffered inside the writer. The writer keeps buffering the
+// string until it detects an EOL character and produces a log message with the
+// buffered string content before the EOL character.
 func NewDebugWriter(l Logger) io.Writer {
-	return &writer{l, DebugString}
+	return &writer{l: l, f: DebugString}
 }
 
 func (w *writer) Write(data []byte) (int, error) {
-	w.f(w.l, string(data))
-	return len(data), nil
+	var n int
+
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	i := bytes.IndexRune(data, '\n')
```

Comment:
```suggestion
	i := bytes.IndexByte(data, '\n')
```

`IndexRune()` requires `data` to be valid UTF-8, but I think we're better off just playing dumb here and splitting anything by newlines.

Reply:
I wonder also if we should be dealing with windows `CRLF`

Reply:
Actually, I am wrong about this as well, we're better off assuming we're getting real text.